onnxscript/rewriter/ort_fusions/attention.py

```diff
@@ -0,0 +1,188 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+from __future__ import annotations
+
+from typing import Sequence, Union
+
+import onnxscript.ir as ir
+from onnxscript.rewriter import pattern
+
+Dim = Union[int, ir.SymbolicDim]
+
+
+# TODO: Maybe add this check to utilities
+def _check_shape(bindings: dict[str, Dim], val: ir.Value, shape: Sequence[str]) -> bool:
+    if val.shape is None:
+        return False
+    if val.shape.rank() != len(shape):
+        return False
+    for actual, expected in zip(val.shape, shape):
+        if expected not in bindings:
+            bindings[expected] = actual  # type: ignore[assignment]
+        elif actual != bindings[expected]:
+            return False
+    return True
+
+
+class Attention(pattern.RewriteRuleClassBase):
```

Comment:
nit: suggest a name like "AttentionFusion".

Another potential issue (unfortunately) is that there is com.microsoft.Attention and onnx Attention (and we may end up needing a fusion for both). But I guess we have a separate folder for ORT contrib op fusion, so I guess that should be enough to distinguish the two.